### PR TITLE
#236 : Changes to file name truncation and checksum calculation

### PR DESCRIPTION
**Pull request by suskumar-MSFT**
* Limited the name truncation logic only to controls
* Found a bug in the check sum comparison, fixed the same

### CODE REVIEW

File: src/PAModel/Serializers/SourceSerializer.cs [hunk 1]
```diff
@@ -950,5 +953,39 @@ private static void AddTestSuiteControl(CanvasDocument app, BlockNode controlIR)
...
+            }
+            return newControlName;
+        }
+
+        /// <summary>
+        /// djb2 algorithm to compute the hash of a string
+        /// </summary>
+        /// <param name="str"></param>
+        /// <returns></returns>
+        private static ulong GetHash(string str)
```
Comment (MikeStall) [inline src/PAModel/Serializers/SourceSerializer.cs:980]: >GetHash [](http://example.com/codeflow?start=29&length=7)

nit: 
addd comment "This must be determinstic and stable since it's used in file names". 
I'd add a unit test to verify that - hash a literal and compare to expected value (like what we do with checksums).

File: src/PAModel/Serializers/SourceSerializer.cs [hunk 2]
```diff
@@ -950,5 +953,39 @@ private static void AddTestSuiteControl(CanvasDocument app, BlockNode controlIR)
...
+        /// We do modulo by 1000 of the hash to limit it to 3 characters.
+        /// </summary>
+        /// <returns></returns>
+        private static string TruncateControlNameIfTooLong(string controlName)
+        {
+            var newControlName = Utilities.EscapeFilename(controlName);
+            if (newControlName.Length > MaxControlNameLength)
+            {
+                // limit the hash to 3 characters by doing a module by 1000
+                var hash = (GetHash(newControlName) % 1000).ToString("x3");
```
Comment (MikeStall) [inline src/PAModel/Serializers/SourceSerializer.cs:969]: >1000 [](http://example.com/codeflow?start=54&length=4)

nit: Since you're printing as hex, could this be 16^3 (instead of 10^3)

File: src/PAModel/Checksum/ChecksumMaker.cs [hunk 3]
```diff
@@ -189,7 +189,8 @@ internal static string ChecksumToString(byte[] bytes)
 
             foreach (var kv in _files.OrderBy(x => x.Key))
             {
-                hash.AppendData(kv.Value);
+                var fileNameInBytes = Encoding.ASCII.GetBytes(kv.Key);
+                hash.AppendData(fileNameInBytes.Concat(kv.Value).ToArray());
```
Comment (MikeStall) [inline src/PAModel/Checksum/ChecksumMaker.cs:193]: Why this change?
This looks like a change to the checksum algorithm, which can be significant (server computes checksum too).

File: src/PAModelTests/FilePathTests.cs [hunk 4]
```diff
@@ -1,27 +0,0 @@
...
-
-using Microsoft.PowerPlatform.Formulas.Tools;
-using Microsoft.PowerPlatform.Formulas.Tools.Utility;
-using Microsoft.VisualStudio.TestTools.UnitTesting;
-using System.IO;
-
-namespace PAModelTests
-{
-    [TestClass]
-    public class FilePathTests
```
Comment (MikeStall) [inline src/PAModelTests/FilePathTests.cs:12]: >FilePathTests [](http://example.com/codeflow?start=17&length=13)

Did you intend to delete this (loss of coverage)?
Is there another test to add to fill the gap?

File: src/PAModel/Checksum/ChecksumMaker.cs (hunk 3 above)
Comment (suskumar-MSFT) [reply src/PAModel/Checksum/ChecksumMaker.cs:193]: we were calculating whole checksum based on the content of the files, not account for the different filenames due to which the issue that my changes caused slipped through. I made this change to calculate the checksum for content of the file along with filename to ensure everything match.

File: src/PAModelTests/FilePathTests.cs (hunk 4 above)
Comment (suskumar-MSFT) [reply src/PAModelTests/FilePathTests.cs:12]: the changes to FilePath were reverted so removed this one, I will add the coverage where it is currently moved to.

File: src/PAModel/Serializers/SourceSerializer.cs (hunk 2 above)
Comment (suskumar-MSFT) [reply src/PAModel/Serializers/SourceSerializer.cs:969]: we are doing the modulo by 1000 before converting to hex, so the max value would 999 in decimal.

File: src/PAModel/Checksum/ChecksumMaker.cs (hunk 3 above)
Comment (MikeStall) [reply src/PAModel/Checksum/ChecksumMaker.cs:193]: I see - we're not including kv.Key in the hash. 

can we just add:
`hash.AppendData(kv.Key);`

and avoid the concat/toArray?

**Comment by MikeStall** [comment]
>         public static string Version = "C6";

If we do change checksum, we should update this to 'C7'

We also have a bunch of tests that will fail (which validate checksum stability). 
This friction is intentional - checksum changes are a big deal. 


---
Refers to: src/PAModel/Checksum/ChecksumMaker.cs:25 in 644f80c. [](commit_id = 644f80c3b32521d8e5fbbcbb582bc69bc5e27b73, deletion_comment = False)

File: src/PAModel/Serializers/SourceSerializer.cs (hunk 2 above)
Comment (MikeStall) [reply src/PAModel/Serializers/SourceSerializer.cs:969]: 999.ToString("x3") // "3e7"
(16*16*16-1).ToString("x3") // "fff"

Comment (MikeStall) [reply src/PAModel/Serializers/SourceSerializer.cs:969]: basically, if the goal is to condense the hash to 3 hex digits, I think you should replace '1000' with (16*16*16-1)

File: src/PAModel/Utility/FilePath.cs [hunk 5]
```diff
@@ -25,19 +25,21 @@ public FilePath(params string[] segments)
 
         public string ToMsAppPath()
         {
-            return string.Join("\\", _pathSegments);
+            var path = string.Join("\\", _pathSegments);
+
+            // Some paths mistakenly start with DirectorySepChar in the msapp,
```
Comment (MikeStall) [inline src/PAModel/Utility/FilePath.cs:30]: >mistakenly [](http://example.com/codeflow?start=26&length=10)

is this a server bug in generating the app?

Comment (suskumar-MSFT) [reply src/PAModel/Utility/FilePath.cs:30]: possibly I have copied over this logic from DirectoryHelper.cs, where we remove this '_' when we load app from sources. @lesaltzm do you know if this was a server bug?

File: src/PAModel/Serializers/SourceSerializer.cs [hunk 6]
```diff
@@ -828,19 +831,19 @@ private static void RestoreViewName(DataSourceEntry ds, string dataSetName)
             }
 
             // Write out of all the other state for roundtripping 
-            string extraContent = (topParentname ?? controlName) + ".editorstate.json";
+            string extraContent = (topParentname ?? newControlName) + ".editorstate.json";
```
Comment (lesaltzm) [inline src/PAModel/Serializers/SourceSerializer.cs:834]: should this always be `newControlName` for editorState?

Comment (suskumar-MSFT) [reply src/PAModel/Serializers/SourceSerializer.cs:834]: we always dump the editorstate data per top control, for AppTest control since we are sharding into individual TestSuites the editorState.json would still be 1.

**Comment by MikeStall** [comment]
Filed https://github.com/microsoft/PowerApps-Language-Tooling/issues/236 to track this.

File: src/PAModel/Checksum/ChecksumMaker.cs [hunk 7]
```diff
@@ -21,8 +21,11 @@ namespace Microsoft.PowerPlatform.Formulas.Tools
     /// </summary>
     public class ChecksumMaker
     {
+        // If the checksum is generated with C6 version then treat it as if it's missing.
+        public static string VersionC6 = "C6";
```
Comment (lesaltzm) [inline src/PAModel/Checksum/ChecksumMaker.cs:25]: Could we make this more generic? Parse the version and if its <= the current one, treat it as if it's missing?

Comment (suskumar-MSFT) [reply src/PAModel/Checksum/ChecksumMaker.cs:25]: do we want to loosen up that much? Earlier we were very strict, we required the exact version

File: src/PAModel/Serializers/MsAppSerializer.cs [hunk 8]
```diff
@@ -227,7 +228,11 @@ public static CanvasDocument Load(Stream streamToMsapp, ErrorContainer errors)
                 // This is debug only. The server checksum is out of date with the client checksum
                 // The main checksum validation that matters is the repack after unpack
 #if DEBUG
```
Comment (MikeStall) [inline src/PAModel/Serializers/MsAppSerializer.cs:230]: Now that we have an "is older" check, can we remove #if Debug and do this all the time?

File: src/PAModel/Serializers/MsAppSerializer.cs [hunk 9]
```diff
@@ -227,7 +228,11 @@ public static CanvasDocument Load(Stream streamToMsapp, ErrorContainer errors)
...
-                if (app._checksum.ServerStampedChecksum != null && app._checksum.ServerStampedChecksum != currentChecksum.wholeChecksum)
+                // In case the server stamped checksum on the msapp is older than the client, treat it like its missing checksum.
+                var isNullOrOlderChecksum = app._checksum.ServerStampedChecksum == null
```
Comment (MikeStall) [inline src/PAModel/Serializers/MsAppSerializer.cs:232]: >app._checksum.ServerStampedChecksum == null [](http://example.com/codeflow?start=44&length=43)

nit: may be cleaner w/ short circuiting.

(app._checksum.ServerStampedChecksum == null) |||
(ChecksumMaker.GetChecksumVersion(app._checksum.ServerStampedChecksum) < ChecksumMaker.Version)

File: src/PAModel/Utility/Utilities.cs [hunk 10]
```diff
@@ -358,5 +360,41 @@ public static string GetResourceRelativePath(ContentKind contentType)
...
+        /// <param name="name">The name that needs to be truncated.</param>
+        /// <returns>Returns the truncated name if the escaped version of it is longer than 50 characters.</returns>
+        public static string TruncateNameIfTooLong(string name)
+        {
+            var escapedName = EscapeFilename(name);
+            if (escapedName.Length > MaxNameLength)
+            {
+                // limit the hash to 3 characters by doing a module by 4096 (16^3)
+                var hash = (GetHash(escapedName) % 4096).ToString("x3");
+                escapedName = escapedName.Substring(0, MaxNameLength - hash.Length) + hash;
```
Comment (MikeStall) [inline src/PAModel/Utility/Utilities.cs:378]: >hash [](http://example.com/codeflow?start=86&length=4)

nit:
Can we inject an underscore, makes it easier to realize it's been truncated and hashed.  Like:

+ "_" + hash;

"FileNameTruncated_fff"

**Review by MikeStall — DISMISSED** [review]
:shipit:

File: src/PAModel/Serializers/MsAppSerializer.cs (hunk 8 above)
Comment (suskumar-MSFT) [reply src/PAModel/Serializers/MsAppSerializer.cs:230]: makes sense
